poly-commitment/src/lagrange_cache.rs

```diff
@@ -0,0 +1,201 @@
+use ark_poly::Radix2EvaluationDomain as D;
+use std::hash::{Hash, Hasher};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LagrangeCache<G> {
+    Local(FileCache<G>),
+    Remote(GoogleCloudCache<G>),
+}
+
+impl<G> Default for LagrangeCache<G> {
+    fn default() -> Self {
+        LagrangeCache::Local(FileCache::new(PathBuf::from("/tmp/lagrange_cache")))
+    }
+}
+
+// Keep this enum private for type safety
+pub enum LagrangeCacheKey {
+    Local(PathBuf),
+    Remote(()),
+}
+
+impl<G: CommitmentCurve> LagrangeCacheTrait<G> for LagrangeCache<G> {
+    type CacheKey = LagrangeCacheKey;
+
+    fn lagrange_basis_cache_key(&self, g: &Vec<G>, domain: &D<G::ScalarField>) -> Self::CacheKey {
+        match self {
+            LagrangeCache::Local(cache) => {
+                LagrangeCacheKey::Local(cache.lagrange_basis_cache_key(g, domain))
+            }
+            LagrangeCache::Remote(cache) => {
+                LagrangeCacheKey::Remote(cache.lagrange_basis_cache_key(g, domain))
+            }
+        }
+    }
+
+    fn load_lagrange_basis_from_cache(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+    ) -> Option<Vec<PolyComm<G>>> {
+        match self {
+            LagrangeCache::Local(cache) => cache.load_lagrange_basis_from_cache(g, domain),
+            LagrangeCache::Remote(cache) => cache.load_lagrange_basis_from_cache(g, domain),
+        }
+    }
+
+    fn cache_lagrange_basis(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+        basis: &Vec<PolyComm<G>>,
+    ) {
+        match self {
+            LagrangeCache::Local(cache) => cache.cache_lagrange_basis(g, domain, basis),
+            LagrangeCache::Remote(cache) => cache.cache_lagrange_basis(g, domain, basis),
+        }
+    }
+}
+
+pub trait LagrangeCacheTrait<G: CommitmentCurve> {
+    type CacheKey;
+
+    fn lagrange_basis_cache_key(&self, g: &Vec<G>, domain: &D<G::ScalarField>) -> Self::CacheKey;
+
+    fn load_lagrange_basis_from_cache(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+    ) -> Option<Vec<PolyComm<G>>>;
+
+    fn cache_lagrange_basis(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+        basis: &Vec<PolyComm<G>>,
+    );
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileCache<G> {
+    cache_dir: PathBuf,
+    point_type: PhantomData<G>,
+}
+
+impl<G> FileCache<G> {
+    pub fn new(cache_dir: PathBuf) -> Self {
+        if !cache_dir.exists() {
+            fs::create_dir_all(&cache_dir).unwrap();
+        }
+        FileCache {
+            cache_dir,
+            point_type: PhantomData,
+        }
+    }
+}
+
+impl<G: CommitmentCurve> LagrangeCacheTrait<G> for FileCache<G> {
+    type CacheKey = PathBuf;
+
+    fn lagrange_basis_cache_key(&self, g: &Vec<G>, domain: &D<G::ScalarField>) -> Self::CacheKey {
+        let mut hasher = DefaultHasher::new();
+        g.hash(&mut hasher);
```

Comment:
Isn't this going to be rather expensive? The lagrange basis has been unchanged for several years, I think it's okay to assume it's constant.

Reply:
@mrmr1993 I noticed from debug printing during tests that the same domain size can appear for multiple different `G`-bases (the `g` field in the `SRS` struct) . Was I hallucinating or is this something that can happen?

Maybe the length of `g` is the only variable, in which case the cache key could be `(g.len(), domain.size())` ?

Reply:
The real variable is `domain.size()` and the curve, we use the same `g`s for all SRSs over the curve (and truncate if the domain doesn't require the full thing).

Reply:
got it :ok_hand:

Reply:
As we discussed, it turns out that for reasons related to chunking we also need to keep the srs size as part of the cache key. Might not even be necessary within the `mina` repo, but it is within `proof-systems`. Still better than hashing the `g` field